doc/designs/quic-design/connection-state-machine.md

```diff
@@ -0,0 +1,633 @@
+  <td rowspan="1"><tt>TERMINATING.DRAINING</tt></td>
+  <td rowspan="1"></td>
+  <td></td>
+  <td></td>
+</tr>
+<tr>
+  <td rowspan="1"><tt>TERMINATED</tt></td>
+  <td rowspan="1"></td>
+  <td>[terminal state]</td>
+  <td></td>
+</tr>
+</table>
+
+Notes on various events:
+
+- `CAN_SEND` is raised when transmission of packets has been unblocked after previously
+  having been blocked. There are broadly two reasons why transmission of packets
+  may not have been possible:
+
+  - Due to OS buffers or network-side write BIOs being full;
+  - Due to limits imposed by the chosen congestion controller.
+
+  `CAN_SEND` is expected to be raised due to a timeout prescribed by the
+  congestion controller or in response to poll(2) or similar notifications, as
+  abstracted by the BIO system and how the application has chosen to notify
+  libssl of network I/O readiness.
+
+  It is generally implied that processing of a packet as mentioned above
+  may cause new packets to be queued and sent, so this is not listed
+  explicitly in the Transition column except for the `CAN_SEND` event.
+
+- `PTO_TIMEOUT` is raised after the PTO interval and stimulates generation
+  of a tail loss probe.
+
+- `IDLE_TIMEOUT` is raised after the connection idle timeout expires.
+  Note that the loss detector only makes a determination of loss due to an
+  incoming ACK frame; if a peer becomes totally unresponsive, this is the only
+  mechanism available to terminate the connection (other than the local
+  application choosing to close it).
+
+- `RX:STATELESS_RESET` indicates receipt of a stateless reset, but note
+  that it is not guaranteed that we are able to recognise a stateless reset
+  that we receive, thus this event may not always be raised.
+
+- `RX:ANY[CONNECTION_CLOSE]` denotes a `CONNECTION_CLOSE` frame received
+  in any non-discarded EL.
+
+- Any circumstance where `RX:RETRY` or `RX:VER_NEG` are not explicitly
+  listed means that these packets are not allowed and will be ignored.
+
+- Protocol errors, etc. can be handled identically to `APP:CLOSE` events
+  as indicated in the above table if locally initiated. Protocol errors
+  signalled by the peer are handled as `RX:ANY[CONNECTION_CLOSE]` events.
+
+Notes on various actions:
+
+- `SendPackets()` sends packets if we have anything pending for transmission,
+  and only to the extent we are able to with regards to congestion control and
+  available BIO buffer space, etc.
+
+Non-FSM Model
+-------------
+
+Common QUIC implementations appear to prefer modelling connection state as a set
+of flags rather than as a FSM. It can be observed above that there is a fair
+degree of commonality between many states. This has been modelled above using
+hierarchical states with default handlers for common events. [The state machine
+can be viewed as a diagram here (large
+image).](./images/connection-state-machine.png)
+
+We transpose the above table to sort by events rather than states, to discern
+the following table:
```

Comment:
It's not a table :grin:
It's also not sorted :disappointed:

Personally, I find this block of text difficult to navigate.